File: ViewModels/BookRideViewModel.swift
//
//  BookRideViewModel.swift
//  UDrive
//
//  Created on 2024
//

import Foundation
import CoreLocation

class BookRideViewModel: ObservableObject {
    @Published var pickupLocation: Location?
    @Published var dropoffLocation: Location?
    @Published var pickupAddress: String = ""
    @Published var dropoffAddress: String = ""
    @Published var isEmergency: Bool = false
    @Published var isLoading: Bool = false
    @Published var errorMessage: String?
    
    private let mockService = MockDataService.shared
    
    func setPickupLocation(_ location: Location) {
        pickupLocation = location
        pickupAddress = location.address
    }
    
    func setDropoffLocation(_ location: Location) {
        dropoffLocation = location
        dropoffAddress = location.address
    }
    
    func bookRide(completion: @escaping (Ride?) -> Void) {
        guard let pickup = pickupLocation, let dropoff = dropoffLocation else {
            errorMessage = "Please select both pickup and dropoff locations"
            return
        }
        
        isLoading = true
        errorMessage = nil
        
        // Mock booking
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) { [weak self] in
            self?.isLoading = false
            let ride = self?.mockService.createRide(pickup: pickup, dropoff: dropoff, isEmergency: self?.isEmergency ?? false)
            completion(ride)
        }
    }
}

File: lib/viewmodels/book_ride_viewmodel.dart
import 'package:flutter/foundation.dart';
import '../services/mock_data_service.dart';
import '../models/ride.dart';
import '../models/location.dart';

class BookRideViewModel extends ChangeNotifier {
  final MockDataService _mockService = MockDataService();
  Location? _pickupLocation;
  Location? _dropoffLocation;
  String _pickupAddress = '';
  String _dropoffAddress = '';
  bool _isEmergency = false;
  bool _isLoading = false;
  String? _errorMessage;

  Location? get pickupLocation => _pickupLocation;
  Location? get dropoffLocation => _dropoffLocation;
  String get pickupAddress => _pickupAddress;
  String get dropoffAddress => _dropoffAddress;
  bool get isEmergency => _isEmergency;
  bool get isLoading => _isLoading;
  String? get errorMessage => _errorMessage;

  set isEmergency(bool value) {
    _isEmergency = value;
    notifyListeners();
  }

  void setPickupAddress(String address) {
    _pickupAddress = address;
    // Find matching GSU location
    final locations = MockDataService.getGSUCampusLocations();
    final location = locations.firstWhere(
      (loc) => loc.address == address,
      orElse: () => Location(
        latitude: 32.5274,
        longitude: -92.7144,
        address: address,
      ),
    );
    _pickupLocation = location;
    notifyListeners();
  }

  void setDropoffAddress(String address) {
    _dropoffAddress = address;
    // Find matching GSU location
    final locations = MockDataService.getGSUCampusLocations();
    final location = locations.firstWhere(
      (loc) => loc.address == address,
      orElse: () => Location(
        latitude: 32.5274,
        longitude: -92.7144,
        address: address,
      ),
    );
    _dropoffLocation = location;
    notifyListeners();
  }

  Future<Ride?> bookRide() async {
    if (_pickupLocation == null || _dropoffLocation == null) {
      _errorMessage = 'Please select both pickup and dropoff locations';
      notifyListeners();
      return null;
    }

    _isLoading = true;
    _errorMessage = null;
    notifyListeners();

    await Future.delayed(const Duration(seconds: 1));

    _isLoading = false;
    final ride = _mockService.createRide(
      pickup: _pickupLocation!,
      dropoff: _dropoffLocation!,
      isEmergency: _isEmergency,
    );
    notifyListeners();
    return ride;
  }
}
